path/a_star.go

```diff
@@ -8,39 +8,34 @@ import (
 
-// HeuristicCoster wraps the HeuristicCost method. A graph that implementing
-// the interface provides a heuristic between any two given nodes.
+// HeuristicCoster wraps the HeuristicCost method. A graph implementing the
+// interface provides a heuristic between any two given nodes.
 type HeuristicCoster interface {
 	HeuristicCost(x, y graph.Node) float64
 }
 
-// Returns an ordered list consisting of the nodes between start and goal. The path will be the
-// shortest path assuming the function heuristicCost is admissible. The second return value is the
-// cost, and the third is the number of nodes expanded while searching (useful info for tuning
-// heuristics). Negative Costs will cause bad things to happen, as well as negative heuristic
-// estimates.
+// AStar returns the A*-shortest path from s to t in g using the heuristic h, and the
+// cost of that path. The path and cost is held in a Shortest along with paths and costs
+// to all nodes explored during the search. The number of expanded nodes is also returned.
+// This value may help with heuristic tuning.
 //
-// A heuristic is admissible if, for any node in the graph, the heuristic estimate of the cost
-// between the node and the goal is less than or set to the true cost.
+// The path will be the shortest path if the heuristic is admissible. A heuristic is
+// admissible if for any node, n, in the graph, the heuristic estimate of the cost of
+// the path from n to t is less than or equal to the true cost of that path.
 //
-// Performance may be improved by providing a consistent heuristic (though one is not needed to
-// find the optimal path), a heuristic is consistent if its value for a given node is less than
-// (or equal to) the actual cost of reaching its neighbors + the heuristic estimate for the
-// neighbor itself. You can force consistency by making your HeuristicCost function return
-// max(NonConsistentHeuristicCost(neighbor,goal), NonConsistentHeuristicCost(self,goal) -
-// Cost(self,neighbor)). If there are multiple neighbors, take the max of all of them.
-//
-// If heuristice is nil, A* will use the graph's HeuristicCost method if present, otherwise
-// falling back to NullHeuristic. To run Uniform Cost Search, run A* with the NullHeuristic.
-//
-// To run Breadth First Search, run A* with both the NullHeuristic and UniformCost (or any cost
-// function that returns a uniform positive value.)
-func AStar(start, goal graph.Node, g graph.Graph, h Heuristic) (path []graph.Node, pathCost float64, nodesExpanded int) {
+// If h is nil, AStar will use the g.HeuristicCost method if g implements HeuristicCoster,
+// falling back to NullHeuristic otherwise. If the graph does not implement graph.Weighter,
+// graph.UniformCost is used. AStar will panic if g has an A*-reachable negative edge weight.
+func AStar(s, t graph.Node, g graph.Graph, h Heuristic) (path Shortest, expanded int) {
+	if !g.Has(s) {
+		return Shortest{from: s}, 0
+	}
```

Comment:
Yes, that's what I meant. But since the docs say "with paths and costs to all nodes explored
during the search", in the end I thought that the caller will get a documented behavior that he asked for. But your solution is better and also in agreement with the docs.